client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -444,44 +347,112 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
     }
 
     /**
-     * Returns an {@link Iterable} of hosts to be used for a request call.
-     * Ideally, the first host is retrieved from the iterable and used successfully for the request.
-     * Otherwise, after each failure the next host has to be retrieved from the iterator so that the request can be retried until
-     * there are no more hosts available to retry against. The maximum total of attempts is equal to the number of hosts in the iterable.
-     * The iterator returned will never be empty. In case there are no healthy hosts available, or dead ones to be be retried,
-     * one dead host gets returned so that it can be retried.
+     * Returns a non-empty {@link Iterator} of nodes to be used for a request
+     * that match the {@link NodeSelector}.
+     * <p>
+     * If there are no living nodes that match the {@link NodeSelector}
+     * this will return the dead node that matches the {@link NodeSelector}
+     * that is closest to being revived.
+     * <p>
+     * If no living and no dead nodes match the selector we retry a few
+     * times to handle concurrent modifications of the list of dead nodes.
+     * We never block the thread or {@link Thread#sleep} or anything like
+     * that. If the retries fail this throws a {@link IOException}.
+     * @throws IOException if no nodes are available
      */
-    private HostTuple<Iterator<HttpHost>> nextHost() {
-        final HostTuple<Set<HttpHost>> hostTuple = this.hostTuple;
-        Collection<HttpHost> nextHosts = Collections.emptySet();
-        do {
-            Set<HttpHost> filteredHosts = new HashSet<>(hostTuple.hosts);
-            for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
-                if (System.nanoTime() - entry.getValue().getDeadUntilNanos() < 0) {
-                    filteredHosts.remove(entry.getKey());
+    private NodeTuple<Iterator<HttpHost>> nextNode(NodeSelector nodeSelector) throws IOException {
+        NodeTuple<List<Node>> nodeTuple = this.nodeTuple;
+        List<HttpHost> hosts = selectHosts(nodeTuple, blacklist, lastNodeIndex, System.nanoTime(), nodeSelector);
+        return new NodeTuple<>(hosts.iterator(), nodeTuple.authCache);
+    }
+
+    static List<HttpHost> selectHosts(NodeTuple<List<Node>> nodeTuple,
+            Map<HttpHost, DeadHostState> blacklist, AtomicInteger lastNodeIndex,
+            long now, NodeSelector nodeSelector) throws IOException {
+        class DeadNodeAndRevival {
```

Comment:
I have some problems looking at this class that is declared within a method :) maybe we should make it a static class and declare it at the main class level?

Reply:
Sure!